File: rotating_the_box_using_two_pointers/rtbutp.py
def rotate_the_box(box_grid: list[list[str]]) -> list[list[str]]:
    m = len(box_grid)
    n = len(box_grid[0])

    for row in range(m):
        empty_col = n - 1

        for col in range(n - 1, -1, -1):
            if box_grid[row][col] == "*":
                empty_col = col - 1
            elif box_grid[row][col] == "#":
                box_grid[row][col] = "."
                box_grid[row][empty_col] = "#"
                empty_col -= 1

    rotated: list[list[str]] = [["." for _ in range(m)] for _ in range(n)]

    for row in range(m):
        for col in range(n):
            rotated[col][m - 1 - row] = box_grid[row][col]

    return rotated


print(rotate_the_box([["#", ".", "#"]]))
print(
    rotate_the_box(
        [
            ["#", ".", "*", "."],
            ["#", "#", "*", "."],
        ]
    )
)

File: rotating_the_box_using_two_pointers/rtbutp.ts
function rotateTheBox(boxGrid: string[][]): string[][] {
    const m = boxGrid.length;
    const n = boxGrid[0].length;

    for (let row = 0; row < m; row++) {
        let emptyCol = n - 1;

        for (let col = n - 1; col >= 0; col--) {
            if (boxGrid[row][col] === '*') {
                emptyCol = col - 1;
            } else if (boxGrid[row][col] === '#') {
                boxGrid[row][col] = '.';
                boxGrid[row][emptyCol] = '#';
                emptyCol--;
            }
        }
    }

    const rotated: string[][] = Array.from({ length: n }, () => Array(m).fill('.'));

    for (let row = 0; row < m; row++) {
        for (let col = 0; col < n; col++) {
            rotated[col][m - 1 - row] = boxGrid[row][col];
        }
    }

    return rotated;
}

console.log(rotateTheBox([['#', '.', '#']]));
console.log(rotateTheBox([
    ['#', '.', '*', '.'],
    ['#', '#', '*', '.'],
]));
